queue_consumer/consumer.py

```diff
@@ -45,40 +50,74 @@ def shutdown(self):
             worker.shutdown()
         support.logger.info('Queue Consumer is shutdown')
 
-    def supervise(self, blocking=False, polling_time=1):
+    def supervise(self,
+                  blocking=False,
+                  polling_time=1,
+                  stuck_time=60,
+                  stuck_limit=None):
+        stuck_limit = stuck_limit or self._max_handlers
         support.logger.debug(
-            f'Queue Consumer supervising in {"blocking" if blocking else "non-blocking"} mode ...')
+            'Queue Consumer supervising in '
+            f'{"blocking" if blocking else "non-blocking"} mode ...')
 
-        def _supervise():
-            while True:
+        args = (polling_time, stuck_time, stuck_limit)
+        if blocking:
+            self._supervise(*args)
+        else:
+            t = Thread(target=self._supervise, args=args, daemon=True)
+            t.start()
+            return t
+
+    def _supervise(self, polling_time, stuck_time, stuck_limit):
+        while True:
+            self._check_workers()
+            self._check_handlers(stuck_time)
 
-                workers = []
-                for worker in self._workers:
+            stuck_handlers_number = len(self._stuck_handlers)
+            if stuck_handlers_number > stuck_limit:
+                raise RuntimeError(
+                    f'Number of stuck handlers {stuck_handlers_number} more '
+                    f'than limit {stuck_limit}')
 
-                    if worker.is_alive():
-                        workers.append(worker)
-                    else:
-                        support.statsd.increment('revived.workers')
+            if self._shutdown:
+                self._no_supervise.set()
+                return
 
-                        new_worker = self._get_worker()
-                        new_worker.start()
-                        workers.append(new_worker)
-                self._workers = workers
+            time.sleep(polling_time)
 
-                if self._shutdown:
-                    self._no_supervise.set()
-                    return
+    def _check_workers(self):
+        workers = []
+        for worker in self._workers:
 
-                time.sleep(polling_time)
+            if worker.is_alive():
+                workers.append(worker)
+            else:
+                support.statsd.increment('revived.workers')
```

Comment:
The only note - I would prefer to increment any metrics / report about any other kind of success or failure after some actions are done. If any of 3 lines below throw an exception and worker is not actually created, metrics are already incremented which might cause an inconsistency between metrics and real state of the application

Reply:
sounds reasonable yeah, let do it